src/Microsoft.NetCore.Analyzers/Core/Security/DoNotReferSelfInSerializableClass.cs

```diff
@@ -0,0 +1,191 @@
+
+                    compilationStartAnalysisContext.RegisterCompilationEndAction(
+                        (CompilationAnalysisContext compilationAnalysisContext) =>
+                        {
+                            ModifyDegree(inDegree, forwardGraph);
+                            ModifyDegree(outDegree, invertedGraph);
+
+                            // If the degree of a vertex is greater than 0 both in the forward graph and inverted graph after topological sorting,
+                            // the vertex must belong to a loop.
+                            var leftVertices = inDegree.Where(s => s.Value > 0).Select(s => s.Key).ToImmutableHashSet();
+                            var invertedLeftVertices = outDegree.Where(s => s.Value > 0).Select(s => s.Key).ToImmutableHashSet();
+                            var verticesInLoop = leftVertices.Intersect(invertedLeftVertices);
+
+                            foreach (var vertex in verticesInLoop)
+                            {
+                                if (vertex is IFieldSymbol)
+                                {
+                                    compilationAnalysisContext.ReportDiagnostic(
+                                        vertex.CreateDiagnostic(Rule, vertex.Name));
+                                }
+                            }
+                        });
+
+                    /// <summary>
+                    /// Add a line to the graph.
+                    /// </summary>
+                    /// <param name="from">The start point of the line</param>
+                    /// <param name="to">The end point of the line</param>
+                    /// <param name="degree">The out degree of all vertices in the graph</param>
+                    /// <param name="graph">The graph</param>
+                    void AddLine(ISymbol from, ISymbol to, ConcurrentDictionary<ISymbol, int> degree, ConcurrentDictionary<ISymbol, HashSet<ISymbol>> graph)
+                    {
+                        var value = AddPoint(from, degree, graph);
+
+                        if (value.Add(to))
+                        {
+                            degree[from]++;
+                        }
+
+                        AddPoint(to, degree, graph);
+                    }
+
+                    /// <summary>
+                    /// Add a point to the graph.
+                    /// </summary>
+                    /// <param name="point">The point to be added</param>
+                    /// <param name="degree">The out degree of all vertices in the graph</param>
+                    /// <param name="graph">The graph</param>
+                    HashSet<ISymbol> AddPoint(ISymbol point, ConcurrentDictionary<ISymbol, int> degree, ConcurrentDictionary<ISymbol, HashSet<ISymbol>> graph)
+                    {
+                        if (!graph.TryGetValue(point, out var value))
+                        {
+                            value = new HashSet<ISymbol>();
```

Comment:
Possible race condition here.  If two threads both enter this if block, then they will return different HashSets.